include/ginkgo/core/preconditioner/ilu.hpp

```diff
@@ -0,0 +1,410 @@
+
+public:
+    static_assert(
+        std::is_same<typename LSolverType::value_type,
+                     typename USolverType::value_type>::value,
+        "Both the L- and the U-solver must use the same `value_type`!");
+    using value_type = typename LSolverType::value_type;
+    using l_solver_type = LSolverType;
+    using u_solver_type = USolverType;
+    static constexpr bool performs_reverse_apply = ReverseApply;
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Factory for the L solver
+         */
+        std::shared_ptr<typename l_solver_type::Factory> GKO_FACTORY_PARAMETER(
+            l_solver_factory, nullptr);
+
+        /**
+         * Factory for the U solver
+         */
+        std::shared_ptr<typename u_solver_type::Factory> GKO_FACTORY_PARAMETER(
+            u_solver_factory, nullptr);
+    };
+
+    /**
+     * Returns the solver which is used for the provided L matrix.
+     *
+     * @returns  the solver which is used for the provided L matrix
+     */
+    std::shared_ptr<const l_solver_type> get_l_solver() const
+    {
+        return l_solver_;
+    }
+
+    /**
+     * Returns the solver which is used for the provided U matrix.
+     *
+     * @returns  the solver which is used for the provided U matrix
+     */
+    std::shared_ptr<const u_solver_type> get_u_solver() const
+    {
+        return u_solver_;
+    }
+
+protected:
+    /**
+     * Manages the `generate` arguments for the parent class to allow multiple
+     * versions to initialize both L and U. Two constructors are provided:
+     * - a Composition, containing the L matrix as the first operand, and the
+     *   U matrix as the second and last (also accepts ParIlu objects)
+     * - both L and U matrix as separate parameters
+     */
+    struct LuArgs {
+        LuArgs(std::shared_ptr<const LinOp> composition)
+        {
+            auto comp_cast =
+                as<const Composition<value_type>>(composition.get());
+            if (comp_cast->get_operators().size() != 2) {
+                throw GKO_NOT_SUPPORTED(comp_cast);
+            }
+            l_factor = comp_cast->get_operators()[0];
+            u_factor = comp_cast->get_operators()[1];
+        }
+
+        LuArgs(std::shared_ptr<const LinOp> l_fac,
+               std::shared_ptr<const LinOp> u_fac)
+            : l_factor{std::move(l_fac)}, u_factor{std::move(u_fac)}
+        {}
+
+        /**
+         * Returns the size that the solver using L and U would return
+         *
+         * @param reverse_apply  determines if the solver solves for U first
+         *                       and then for L (reverse_apply = true), or
+         *                       first with L, then with U
+         *                       (reverse_apply = false)
+         *
+         * @returns the size that the solver using L and U would return
+         */
+        dim<2> get_solver_size(bool reverse_apply = false) const
```

Comment:
Should this be a parameter? Since you already have the reverse apply templated parameter?

Reply:
You are right, the parameter is unnecessary. I removed it in the latest version.
Thanks!